miasm2/core/asmblock.py

```diff
@@ -359,132 +323,209 @@ class AsmSymbolPool(object):
     """
 
     def __init__(self):
-        self._labels = set()
-        self._name2label = {}
-        self._offset2label = {}
-        self._label_num = 0
+        self._loc_keys = set()
+
+        self._loc_key_to_offset = {}
+        self._loc_key_to_name = {}
+
+        self._name_to_loc_key = {}
+        self._offset_to_loc_key = {}
+
+        self._loc_key_num = 0
 
-    def add_label(self, name, offset=None):
+    def loc_key_to_offset(self, loc_key):
         """
-        Create and add a label to the symbol_pool
-        @name: label's name
-        @offset: (optional) label's offset
+        Return offset of @loc_key, None else.
+        @loc_key: LocKey instance
         """
-        label = AsmLabel(name, offset)
+        return self._loc_key_to_offset.get(loc_key, None)
+
+    def loc_key_to_name(self, loc_key):
+        """
+        Return name of @loc_key.
+        @loc_key: LocKey instance
+        """
+        return self._loc_key_to_name.get(loc_key)
+
+    def add_location(self, name, offset=None):
+        """
+        Create and add a location to the symbol_pool
+        @name: loc_key's name (never empty)
+        @offset: (optional) loc_key's offset
+        """
+
+        if is_int(name):
+            assert offset is None or offset == name
+            offset = name
+            name = "loc_%.16X" % (int(name) & 0xFFFFFFFFFFFFFFFF)
+        if offset is not None:
+            offset = int(offset)
+
+        assert name
+        assert not is_int(name)
 
         # Test for collisions
-        if (label.offset in self._offset2label and
-                label != self._offset2label[label.offset]):
-            raise ValueError('symbol %s has same offset as %s' %
-                             (label, self._offset2label[label.offset]))
-        if (label.name in self._name2label and
-                label != self._name2label[label.name]):
-            raise ValueError('symbol %s has same name as %s' %
-                             (label, self._name2label[label.name]))
-
-        self._labels.add(label)
-        if label.offset is not None:
-            self._offset2label[label.offset] = label
-        if label.name != "":
-            self._name2label[label.name] = label
-        return label
-
-    def remove_label(self, label):
+        known_loc_key = self.getby_name(name)
+        if known_loc_key is not None:
+            known_offset = self.loc_key_to_offset(known_loc_key)
```

Comment:
`known_offset` has to be checked against `None`, and then `known_loc_key` associated to `offset` argument.
Use case:

Already in symbol_pool: (loc_key, name, offset) 1, "xx", None

`add_location("xx", 14)`
Expected output is, IMHO, one entry with `1, "xx", 14`

Reply:
The following `if` (L.379) statements can maybe merged with this new condition

Reply:
not a bug